Домашка. Week 2

anton0701 opened the pull request:
Сделал обязательную часть 
Из опционального сделал 2 пункта:
[опционально] Создать несколько конфигурационных файлов для разных сред для обоих проектов
[опционально] Написать парсер конфигов для обоих проектов

Постараюсь доделать опциональные пункты, пока не успеваю

Из того что не просили, но я сделал: поставил zap logger + добавил проверки полей для запросов и возвращаю ошибки, если проверки не проходят

P.S. postgres/cmd/query + postgres/cmd/query_with_squirel: файлы не используются -- просто повторял за Олегом, чтобы руку набить, вприницпе можно удалить

anton0701 committed:
postgres + docker-compose + migrations added

anton0701 committed:
CRUD implemented

anton0701 committed:
добавил dbPool как проперти server, чтобы не пересоздавать пул соединений к бд при каждом запросе

anton0701 committed:
docker-compose: нейминг сервиса и volume (тома)

anton0701 committed:
минор

anton0701 committed:
zap logger added

anton0701 committed:
zap logger added

anton0701 committed:
добавлена проверка полей в запросе

anton0701 committed:
adding local + prod configs. beginning

anton0701 committed:
makefile: доавлены миграции для прод окружения

Shemistan commented on grpc/grpc_server/main.go:
```diff
@@ -2,75 +2,290 @@ package main
 type server struct {
 	desc.UnimplementedUserV1Server
+	dbPool *pgxpool.Pool
+	log    *zap.Logger
```
супер

anton0701 replied:
++

Shemistan commented on grpc/grpc_server/main.go:
```diff
@@ -2,75 +2,290 @@ package main
+
+	// TODO: правильно?
+	var updatedAtProto *timestamppb.Timestamp
+	if updatedAt.Valid {
```
Лучше это поле не ожидать из вне, а генерить на уровне сервиса
А в сервисе пользоваться уже time.time

anton0701 replied:
Здесь и в строках кода 143-146 не до конца понял комментарий.

Напишу свои пояснения относительно кода.

Это код метода Get-User. 
Переменная updatedAt - это переменная типа sql.NullTime, значение которой получаю из запроса к БД, колонка updated_at. 
Думаю в методе Get выставлять значение updated_at -- неправильно по смыслу, тк Get ничего не апдейтит.

В методе (s *server) UpdateUser(ctx, req) я меняю значение updated_at, выставляя значение time.Now()

Если я неправильно понял твой комментарий, пожалуйста, переформулируй коммент)

anton0701 replied:
++

Shemistan commented on grpc/grpc_server/main.go:
```diff
@@ -2,75 +2,290 @@ package main
+	if updatedAt.Valid {
+		updatedAtProto = timestamppb.New(updatedAt.Time)
+	} else {
+		updatedAtProto = nil
```
Лишнее, он и так нил будет. Но писал выше - лучше это выставлять самому в сервисе7
Не приятно увидеть в БД - что кто-то отредактировал данные и не понятно когда это произошло

anton0701 replied:
см. коммент выше ("Здесь и в строках кода 143-146 не до конца понял....")

ссылка на коммент выше -- https://github.com/anton0701/auth/pull/2#discussion_r1695593428

anton0701 replied:
По поводу лишеного if -- убрал

Shemistan commented on grpc/grpc_server/main.go:
```diff
@@ -2,75 +2,290 @@ package main
+		return nil, status.Errorf(codes.Internal, "Unable to create SQL query from builder. Error info: %v", err)
+	}
+
+	var (
```
Это ок, пока до архитектуры не дойдем

anton0701 replied:
в следующей домашке переделаю

Shemistan commented on grpc/grpc_server/main.go:
```diff
@@ -2,75 +2,290 @@ package main
+	err = s.dbPool.
+		QueryRow(ctx, query, args...).
+		Scan(&id, &name, &email, &role, &createdAt, &updatedAt)
+
```
лишний пропуск
Обычно хорошей практикой считается объединение вызова и обработки ошибки в один блок кода без пустой строки между ними, чтобы визуально подчеркнуть их связь.

anton0701 replied:
++

Shemistan commented on grpc/grpc_server/main.go:
```diff
@@ -2,75 +2,290 @@ package main
 
-func (s *server) GetUserInfo(_ context.Context, req *desc.GetUserInfoRequest) (*desc.GetUserInfoResponse, error) {
-	log.Printf("%s\nMethod Get.\nInput params:\n%+v\n************\n\n", grpcUserAPIDesc, req)
+func (s *server) GetUserInfo(ctx context.Context, req *desc.GetUserInfoRequest) (*desc.GetUserInfoResponse, error) {
```
Для публичных методов и сущностей принято писать описание, например // Create - создает нового пользователя. Такие подсказки например выводятся при наведении на какой-то метод и в принципе помогает лучше понимать код

anton0701 replied:
++

Shemistan commented on grpc/grpc_server/main.go:
```diff
@@ -2,75 +2,290 @@ package main
 
-func (s *server) CreateUser(_ context.Context, req *desc.CreateUserRequest) (*desc.CreateUserResponse, error) {
-	log.Printf("%s\nMethod Create.\nInput params:\n%+v\n************\n\n", grpcUserAPIDesc, req)
+func (s *server) CreateUser(ctx context.Context, req *desc.CreateUserRequest) (*desc.CreateUserResponse, error) {
```
Описание

anton0701 replied:
++

Shemistan commented on grpc/grpc_server/main.go:
```diff
@@ -2,75 +2,290 @@ package main
+func (s *server) CreateUser(ctx context.Context, req *desc.CreateUserRequest) (*desc.CreateUserResponse, error) {
+	s.log.Info("Method Create-User", zap.Any("Input params", req))
+
+	// TODO: текст ошибки вынести в константу?
```
Когда дойдем до слоистой архитектуры то да лучше выностить все структуры в model/model.go а ошибки можно в service_error/error.go

anton0701 replied:
доделаю в следующей домашке -- про слоистую архитектуру

Shemistan commented on grpc/grpc_server/main.go:
```diff
@@ -2,75 +2,290 @@ package main
 
-func (s *server) UpdateUser(_ context.Context, req *desc.UpdateUserRequest) (*emptypb.Empty, error) {
-	log.Printf("%s\nMethod Update.\nInput params:\n%+v\n************\n\n", grpcUserAPIDesc, req)
+func (s *server) UpdateUser(ctx context.Context, req *desc.UpdateUserRequest) (*emptypb.Empty, error) {
```
Описание

anton0701 replied:
++

Shemistan commented on postgres/Makefile:
```diff
@@ -0,0 +1,29 @@
+local-migration-down:
+	${LOCAL_BIN}/goose -dir $(LOCAL_MIGRATION_DIR) postgres $(LOCAL_MIGRATION_DSN) down -v
+
+# Коряво, но работает:)
```
хорошо

anton0701 replied:
++

anton0701 committed:
добавлены описания методов

anton0701 committed:
добавлены описания публичных методов, интерфейсов и структур

anton0701 committed:
добавлены описания публичных методов, интерфейсов и структур

anton0701 committed:
добавлена валидация запросов:
- интерфейс Validator с функцией Validate()
- методы Validate(), где ресиверами являются структуры реквестов

anton0701 committed:
добавил проверку, что запросы имплементируют интерфейс Validator

anton0701 committed:
убрал лишний if
